### PR TITLE
Java: Using Custom Approval template with old spring security oauth library can lead to remote code execution

### PR DESCRIPTION
**Pull request by dogangcr**
This rule is written for CVE-2018-1260.
Spring Security OAuth, versions 2.3 prior to 2.3.3 and 2.2 prior to 2.2.2 and 2.1 prior to 2.1.2 and 2.0 prior to 2.0.15 and older unsupported versions, contains a remote code execution vulnerability. A malicious user or attacker can craft an authorization request to the authorization endpoint that can lead to a remote code execution when the resource owner is forwarded to the approval endpoint.

This vulnerability exposes applications that meet all of the following requirements:

- Act in the role of an Authorization Server (e.g. @EnableAuthorizationServer)

- Use the default Approval Endpoint

Spring security has OAuth solution for solve single sign on requirements. If developers use this library with old spring security oauth library and default approval page , developer will be affected from this vulnerability. 

OAuth authorization code flow is like below. CVE-2018-1260 is focused on  step 4. RCE attacks happened in step 4, if developer used default approval page. For prevent it , custom approval page or latest jar should be used.
image

This rule finds functions that use act in the role of an authorization server (e.g. @EnableAuthorizationServer) and default the default approval endpoint(forward:/oauth/confirm_access).
Also checks the pom.xml for specific jar version. Versions 2.3 prior to 2.3.3 and 2.2 prior to 2.2.2 and 2.1 prior to 2.1.2 and 2.0 prior to 2.0.15 and older are affected from this vulnerability.
If this version is from the vulnerable set, rule will display it.

### CODE REVIEW

File: java/ql/src/Frameworks/Spring/XML Configuration Errors/insecureOAuthVersion.ql [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from Annotation ann, MethodAccess call, StringLiteral arg, PomVersionChecker config
+where
+  ann
+      .getType()
+      .hasQualifiedName("org.springframework.security.oauth2.config.annotation.web.configuration",
+        "EnableAuthorizationServer") and
+  call.getMethod().hasName("setUserApprovalPage") and
+  call.getArgument(0) = arg and
+  arg.getValue() = "forward:/oauth2/confirm_access" and
+  config.getLibraryData() and  config.isVersionNumberVulnerable()
```
Comment (JLLeitschuh) [inline java/ql/src/Frameworks/Spring/XML Configuration Errors/insecureOAuthVersion.ql]: 😢 Unfortunately, this means that you won't be able to check Gradle based projects for this vulnerability.
Idk if there's any way to get around that though.

File: java/ql/src/Frameworks/Spring/XML Configuration Errors/insecureOAuthVersion.ql [hunk 2]
```diff
@@ -0,0 +1,44 @@
+/**
+ * @name CVE-2018-1260: Remote Code Execution with spring-security-oauth2
+ * @description A malicious user or attacker can craft an authorization request to the authorization endpoint that can lead to a remote code execution when the resource owner is forwarded to the approval endpoint.
+ * @kind Remote Code Execution
+ */
```
Comment (JLLeitschuh) [inline java/ql/src/Frameworks/Spring/XML Configuration Errors/insecureOAuthVersion.ql]: You're going to want to look at this (and potentially other example queries) for how to format this comment correctly.

https://help.semmle.com/QL/learn-ql/writing-queries/query-metadata.html

Comment (dogangcr) [reply java/ql/src/Frameworks/Spring/XML Configuration Errors/insecureOAuthVersion.ql]: Thanks for the link. I'll check it .

File: java/ql/src/Frameworks/Spring/XML Configuration Errors/insecureOAuthVersion.ql (hunk 1 above)
Comment (dogangcr) [reply java/ql/src/Frameworks/Spring/XML Configuration Errors/insecureOAuthVersion.ql]: Is there any way to look for build.gradle files with ql? If you know it, please let me know.

Comment (JLLeitschuh) [reply java/ql/src/Frameworks/Spring/XML Configuration Errors/insecureOAuthVersion.ql]: None thai I know of.

**Comment by aschackmull** [comment]
While this is a nice query (and thank you for your contribution!), this is the sort of thing that's much better handled by dependabot, which in fact already handles this in a much more general way. QL queries are better suited for identifying a particular pattern that makes up a vulnerability in itself. As such we'll close this as being out-of-scope.

**Pull request closed without merging**